### PR TITLE
*: abstract the progress channel (updateCh) into the glue package

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Integrating BR into TiDB requires the caller to know the current progress in order to report it in the SQL result. The current implementation only supports either printing to log or stdout, which is insufficient for the purpose.

### What is changed and how it works?

Move the `utils.StartProgress` function into the `Glue`, so that the TiDB integration can provide its own progress reporting mechanism.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change
    * New method in `glue.Glue`.

Side effects

Related changes

### CODE REVIEW

**Comment by kennytm** [comment]
https://internal.pingcap.net/idc-jenkins/blue/rest/organizations/jenkins/pipelines/br_ghpr_unit_and_integration_test/runs/1027/log/?start=0 🤔 

```
[2020-03-18T19:48:03.745Z] panic: runtime error: slice bounds out of range [:-8]
[2020-03-18T19:48:03.745Z] 
[2020-03-18T19:48:03.745Z] goroutine 614 [running]:
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/restore.truncateTS(...)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/util.go:313
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/restore.(*FileImporter).downloadSST(0xc00020b578, 0xc000645500, 0xc00106b800, 0xc0009ad620, 0x2d414ef, 0xc, 0x1008d96)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/import.go:355 +0xa55
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/restore.(*FileImporter).Import.func1.1(0xc0009f1b00, 0x8)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/import.go:217 +0xcc
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/utils.WithRetry(0x31d7ac0, 0xc000bd68c0, 0xc00177d3a8, 0x31ab980, 0xc0009f1b00, 0x1b, 0x1b)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/utils/retry.go:29 +0x8e
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/restore.(*FileImporter).Import.func1(0x0, 0x0)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/import.go:212 +0x4f6
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/utils.WithRetry(0x31d7ac0, 0xc000bd68c0, 0xc00177de30, 0x31ab980, 0xc001201080, 0xc001201040, 0x1b)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/utils/retry.go:29 +0x8e
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/restore.(*FileImporter).Import(0xc00020b578, 0xc00106b800, 0xc0009ad620, 0xc001132a80, 0x1)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/import.go:197 +0x687
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/restore.(*Client).RestoreFiles.func2()
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/client.go:480 +0xfc
[2020-03-18T19:48:03.745Z] github.com/pingcap/br/pkg/utils.(*WorkerPool).Apply.func1(0xc001698540, 0xc0006f2500, 0xc0005484d0)
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/utils/worker.go:47 +0x31
[2020-03-18T19:48:03.745Z] created by github.com/pingcap/br/pkg/utils.(*WorkerPool).Apply
[2020-03-18T19:48:03.745Z] 	/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/utils/worker.go:46 +0xac
```

File: pkg/task/restore.go [hunk 1]
```diff
@@ -400,8 +400,9 @@ func RunRestoreTiflashReplica(c context.Context, g glue.Glue, cmdName string, cf
 	for _, db := range dbs {
 		tables = append(tables, db.Tables...)
 	}
-	updateCh := utils.StartProgress(
+	updateCh := g.StartProgress(
 		ctx, "RecoverTiflashReplica", int64(len(tables)), !cfg.LogProgress)
+	defer updateCh.Close()
```
Comment (overvenus) [inline pkg/task/restore.go]: `defer close` always make progress to 100% even if it fails. Could you close it only if there is no error?

**Comment by kennytm** [comment]
/run-integration-test pd=release-4.0 tikv=release-4.0 tidb=release-4.0

File: pkg/restore/util.go [hunk 2]
```diff
@@ -309,6 +310,9 @@ func matchNewPrefix(key []byte, rewriteRules *RewriteRules) *import_sstpb.Rewrit
 }
 
 func truncateTS(key []byte) []byte {
+	if len(key) == 0 {
```
Comment (3pointer) [inline pkg/restore/util.go]: Did this fix `panic: runtime error: slice bounds out of range [:-8]`?

Comment (kennytm) [reply pkg/restore/util.go]: This fix the panic but it seems the `key` being nil itself is a bug. We can remove this change.

**Comment by 3pointer** [comment]
/run-integration-test pd=release-4.0 tikv=release-4.0 tidb=release-4.0

**Pull request merged**